sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+        """Validates the input arguments. """
+
+        # Check n_clusters
+        if (self.n_clusters is None or self.n_clusters <= 0 or
+                not isinstance(self.n_clusters, (int, np.integer))):
+            raise ValueError("n_clusters should be a nonnegative integer. "
+                             "%s was given" % self.n_clusters)
+
+        # Check init
+        init_methods = ['random', 'heuristic']
+        if self.init not in init_methods:
+            raise ValueError("init needs to be one of " +
+                             "the following: " +
+                             "%s" % init_methods)
+
+        # Check random state
+        self.random_state_ = check_random_state(self.random_state)
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features).
+            Dataset to cluster.
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        if self.n_clusters > X.shape[0]:
+            raise ValueError("The number of medoids %d must be less "
+                             "than the number of samples %d."
+                             % (self.n_clusters, X.shape[0]))
+
+        distances = pairwise_distances(X, metric=self.distance_metric)
+
+        medoid_idxs = self._get_initial_medoid_indices(distances,
+                                                       self.n_clusters)
+        labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+        # Old medoids will be stored here for reference
+        old_medoid_idxs = np.zeros((self.n_clusters,))
+
+        # Continue the algorithm as long as
+        # the medoids keep changing and the maximum number
+        # of iterations is not exceeded
+        self.n_iter_ = 0
+        while (not np.all(old_medoid_idxs == medoid_idxs) and
+                self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            old_medoid_idxs = np.copy(medoid_idxs)
+            labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_idxs_in_place(distances, labels, medoid_idxs)
+
+        # Expose labels_ which are the assignments of
+        # the training data to clusters
+        self.labels_ = labels
+        self.cluster_centers_ = X[medoid_idxs]
+        self.inertia_ = self._compute_inertia(X)
+
+        # Return self to enable method chaining
+        return self
+
+    def _update_medoid_idxs_in_place(self, distances, cluster_idxs,
+                                     medoid_idxs):
+        """In-place update of the medoid indices"""
+
+        # Update the medoids for each cluster
+        for k in range(self.n_clusters):
+
+            if sum(cluster_idxs == k) == 0:
+                warnings.warn("Cluster %d is empty!" % k)
```

Comment:
When does this happen? When multiple medoids are identical? Will this possibly resolve by itself? Is it something that we should warn about after clustering is finished (by checking `np.all(bincount(labels, minlength=self.n_clusters))` and issuing a `ConvergenceWarning`) rather than at each iteration? Or is the idea of this warning to suggest that the user may want to stop clustering?

I think you also should note somewhere that in this case of non-convergence, the `medoid_idx` may not be labelled with its own cluster.

Reply:
That's an interesting question, I'll try to setup a test and see if this is indeed possible at the end of the fit.

Reply:
So I got a situation that demonstrates this situation:

    >>> from sklearn.cluster import KMedoids
    >>> X = [[1],[1],[1]]
    >>> kmedoids = KMedoids(n_clusters=2).fit(X)
    C:\Users\zane dufour\scikit-learn\sklearn\cluster\k_medoids_.py:196: UserWarning: Cluster 1 is empty!
      warnings.warn("Cluster %d is empty!" % k)
    >>> print(kmedoids.labels_)
    [0 0 0]
    >>> print(kmedoids.medoid_indices_) # I added medoid_indices_ as an instance variable.
    [0 1]

@jnothman Where do you think that should be noted? It seems like something that should be transparent to the user, but I'm not sure where in the documentation it should go.